athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchHelper.java

```diff
@@ -0,0 +1,734 @@
+            // NOTE: Gets called at construction and each call to doListSchemaNames() when autoDiscoverEndpoint is true.
+
+            // AWS ES Client for retrieving domain mapping info from the Amazon Elasticsearch Service.
+            AWSElasticsearch awsEsClient = AWSElasticsearchClientBuilder.defaultClient();
+
+            try {
+                ListDomainNamesResult listDomainNamesResult = awsEsClient.listDomainNames(new ListDomainNamesRequest());
+                List<String> domainNames = new ArrayList<>();
+                for (DomainInfo domainInfo : listDomainNamesResult.getDomainNames()) {
+                    domainNames.add(domainInfo.getDomainName());
+                }
+
+                DescribeElasticsearchDomainsRequest describeDomainsRequest = new DescribeElasticsearchDomainsRequest();
+                describeDomainsRequest.setDomainNames(domainNames);
+                DescribeElasticsearchDomainsResult describeDomainsResult =
+                        awsEsClient.describeElasticsearchDomains(describeDomainsRequest);
+
+                setDomainMapping(describeDomainsResult.getDomainStatusList());
+            }
+            catch (Exception error) {
+                logger.error("Error getting list of domain names:", error);
+            }
+            finally {
+                awsEsClient.shutdown();
+            }
+        }
+        else {
+            // Get domain mapping from environment variable.
+            String domainMapping = getEnv(DOMAIN_MAPPING);
+            if (!domainMapping.isEmpty()) {
+                setDomainMapping(domainSplitter.split(domainMapping));
+            }
+        }
+
+        logger.info("setDomainMapping - exit");
+    }
+
+    /**
+     * Gets a list of domain names.
+     * @return a list of domain-names stored in the domainMap.
+     */
+    public static final Set<String> getDomainList()
+    {
+        return domainMap.keySet();
+    }
+
+    /**
+     * Gets the domain's endpoint.
+     * @param domainName is the name associated with the endpoint.
+     * @return a String containing the endpoint associated with the domainName, or an empty String if the domain-
+     * name doesn't exist in the map.
+     */
+    public static final String getDomainEndpoint(String domainName)
+    {
+        String endpoint = "";
+
+        if (domainMap.containsKey(domainName)) {
+            endpoint = domainMap.get(domainName);
+        }
+
+        return endpoint;
+    }
+
+    /**
+     * Allows for coercion of field-value types in the event that the returned type does not match the schema.
+     * Multiple fields in Elasticsearch can be returned as a string, numeric (Integer, Long, Double), or null.
+     * @param field is the field that we are coercing the value into.
+     * @param fieldValue is the value to coerce
+     * @return the coerced value.
+     */
+    public static Object coerceField(Field field, Object fieldValue)
```

Comment:
Is there a reason we did not use the 'extractor' framework that was used in the example: https://github.com/awslabs/aws-athena-query-federation/blob/master/athena-example/src/main/java/com/amazonaws/connectors/athena/example/ExampleRecordHandler.java#L153

This coerceField method is doing lots of repeated/extra work to translate every records from the source. The extractor framework allows you too greatly reduce the 'branching' and other overhead associated with extracting the value from the source result into ApacheArrow.

Reply:
Added field extractors.